collector/framework_collector.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2016 Mesosphere, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/dcos/dcos-metrics/producers"
+	"github.com/linkedin/goavro"
+)
+
+var (
+	listenEndpointFlag         = "127.0.0.1:8124"
+	recordInputLogFlag         = false
+	recordInputHexdumpFlag     = false
+	inputLimitAmountKBytesFlag = 20480
+	inputLimitPeriodFlag       = 60
+)
+
+// AvroDatum{} conveys the avro data coming in on our
+// TCP listening channel
+type AvroDatum struct {
+	// *goavro.Record
+	Record interface{}
+	// the topic that the Record requested
+	Topic string
+	// the approximate byte size of the original Record, if known
+	ApproxBytes int64
+}
+
+// avroRecord{} conveys field for goavro.Record
+// schema set by the schema package
+type avroRecord []struct {
+	Name   string `json"name"`
+	Fields []struct {
+		Name  interface{} `json:"name"`
+		Datum interface{} `json:"datum"`
+	} `json:"fields"`
+}
+
+// avroRecord.extract() gets tags and datapoints from avro formatted data
+// and adds them to MetricsMessage{}
+func (ar avroRecord) extract(pmm *producers.MetricsMessage) error {
+	fieldType := ""
+	if len(ar) > 0 {
+		fieldType = ar[0].Name
+	} else {
+		return errors.New("No records found for extract.")
+	}
+
+	if fieldType == "dcos.metrics.Tag" {
+		labels := make(map[string]string)
+		for _, field := range ar {
+			log.Debugf("Adding tag %s", field)
+			labels[fmt.Sprintf("%s", field.Fields[0].Datum)] = fmt.Sprintf("%s", field.Fields[1].Datum)
+		}
+		pmm.Dimensions.Labels = labels
+	}
+
+	if fieldType == "dcos.metrics.Datapoint" {
+		datapoints := []producers.Datapoint{}
+		for _, field := range ar {
+			log.Debugf("Adding datapoint %s", field)
+			dp := producers.Datapoint{
+				Name:  fmt.Sprintf("%s", field.Fields[0].Datum),
+				Value: fmt.Sprintf("%s", field.Fields[1].Datum),
+				Unit:  fmt.Sprintf("%s", field.Fields[2].Datum),
+			}
+			datapoints = append(datapoints, dp)
+		}
+		pmm.Datapoints = datapoints
+	}
+
+	if fieldType == "" {
```

Comment:
maybe we can use `else if` here?
it looks like we are checking `fieldType` for empty value or `dcos.metrics.Tag` and `dcos.metrics.Datapoint`
what if we received something else?